internal/storage/groups.go

```diff
@@ -0,0 +1,391 @@
+
+	rows, err := s.db.Statement(ctx).
+		Delete("group_members").
+		Where(sq.Eq{"group_id": groupID}).
+		Suffix("RETURNING user_id").
+		QueryContext(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("failed to remove all users from group: %v", err)
+	}
+	defer rows.Close()
+
+	userIDs := make([]string, 0)
+	for rows.Next() {
+		var userID string
+		if err := rows.Scan(&userID); err != nil {
+			return nil, fmt.Errorf("failed to scan user ID: %v", err)
+		}
+		userIDs = append(userIDs, userID)
+	}
+
+	if err := rows.Err(); err != nil {
+		return nil, fmt.Errorf("error iterating removed users: %v", err)
+	}
+
+	return userIDs, nil
+}
+
+// GetGroupsForUser retrieves all groups that a user belongs to.
+func (s *Storage) GetGroupsForUser(ctx context.Context, userID string) ([]*types.Group, error) {
+	ctx, span := s.tracer.Start(ctx, "storage.Storage.GetGroupsForUser")
+	defer span.End()
+
+	rows, err := s.db.Statement(ctx).
+		Select("g.id", "g.name", "g.tenant_id", "g.description", "g.type", "g.created_at", "g.updated_at").
+		From("groups g").
+		Join("group_members gm ON g.id = gm.group_id").
+		Where(sq.Eq{"gm.user_id": userID}).
+		OrderBy("g.name ASC").
+		QueryContext(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("failed to query groups for user: %v", err)
+	}
+	defer rows.Close()
+
+	groups := make([]*types.Group, 0)
+	for rows.Next() {
+		group, err := scanGroup(rows)
+		if err != nil {
+			return nil, fmt.Errorf("failed to scan group: %v", err)
+		}
+		groups = append(groups, group)
+	}
+
+	if err := rows.Err(); err != nil {
+		return nil, fmt.Errorf("error iterating groups: %v", err)
+	}
+
+	return groups, nil
+}
+
+// UpdateGroupsForUser replaces all group memberships for a user with the specified groups.
+func (s *Storage) UpdateGroupsForUser(ctx context.Context, userID string, groupIDs []string) error {
+	ctx, span := s.tracer.Start(ctx, "storage.Storage.UpdateGroupsForUser")
+	defer span.End()
+
+	if len(groupIDs) == 0 {
+		_, err := s.db.Statement(ctx).
+			Delete("group_members").
+			Where(sq.Eq{"user_id": userID}).
+			ExecContext(ctx)
+		if err != nil {
+			return fmt.Errorf("failed to remove all group memberships: %v", err)
+		}
+		return nil
+	}
+
+	now := time.Now()
+
+	insert := s.db.Statement(ctx).
+		Insert("group_members").
+		Columns("group_id", "user_id", "tenant_id", "role", "created_at", "updated_at").
+		Suffix("ON CONFLICT (group_id, user_id) DO UPDATE SET updated_at = EXCLUDED.updated_at")
+
+	for _, groupID := range groupIDs {
+		insert = insert.Values(groupID, userID, "default", "member", now, now)
+	}
+
+	if _, err := insert.ExecContext(ctx); err != nil {
+		if IsForeignKeyViolation(err) {
+			return WrapForeignKeyError(err, "one or more groups do not exist")
+		}
+		return fmt.Errorf("failed to upsert group memberships: %v", err)
+	}
+
+	_, err := s.db.Statement(ctx).
+		Delete("group_members").
+		Where(sq.And{
+			sq.Eq{"user_id": userID},
+			sq.NotEq{"group_id": groupIDs},
+		}).
+		ExecContext(ctx)
```

Comment:
add explanation of what's happening

in my mind first we clean and then we add, not the opposite OR we do a diff and leave the suff that is supposed to stay as is and delete the rest

simple explanation should suffice and leave the logic as is

Reply:
I also would expect a clean-first add-second approach but as long as the whole transaction is rolled back in case of failure for any of the 2 queries, I'm good